doc/jass.1
.\"	Copyright (c) 2013 Twitter, Inc.
.\"	Originally written by Jan Schaumann <jschauma@twitter.com> in
.\"	April 2013.
.\"
.Dd May 21, 2022
.Dt JASS 1
.Os
.Sh NAME
.Nm jass
.Nd just another secret sharer
.Sh SYNOPSIS
.Nm
.Op Fl Vdehlv
.Op Fl c Ar cmd
.Op Fl f Ar file
.Op Fl g Ar group
.Op Fl k Ar key
.Op Fl p Ar passin
.Op Fl u Ar user
.Sh DESCRIPTION
The
.Nm
tool allows you to share a secret with other people through the use of SSH
keypairs.
It accepts input on stdin and generates ciphertext on stdout, encrypted
for the given key or user.
.Sh OPTIONS
The following options are supported by
.Nm :
.Bl -tag -width _p_passin_
.It Fl V
Print version number and exit.
.It Fl c Ar cmd
Use this command to retrieve the public key.
See
.Xr "FINDING PUBKEYS"
for details.
.It Fl d
Perform decryption.
.It Fl e
Perform encryption.
This is the default.
.It Fl f Ar file
Encrypt/decrypt this file.
If not specified,
.Nm
will read data from stdin.
.It Fl g Ar group
Encrypt for all members of this group.
.It Fl h
Print a short help message and exit.
.It Fl k Ar key
Encrypt or decrypt the input using the key(s) found in
.Ar key .
.It Fl l
List recipients of the encrypted input.
.It Fl p Ar passin
Specify the method to retrieve the passphrase for the private key.
See
.Xr "PASS PHRASE ARGUMENTS"
for details.
Defaults to interactively prompt on the controlling tty.
.It Fl u Ar user
Encrypt the input for this user, attempting to find suitable public SSH
keys on the local system or from LDAP or the enabled URLs.
.It Fl v
Be verbose.
Can be specified multiple times.
.El
.Sh DETAILS
It is not uncommon for people within a given organization to have a need
to share secrets with one another.
.Nm
lets you do this through the use of SSH RSA keys, as those are frequently
already well distributed and trusted.
.Pp
Since asymmetric encryption is only suitable for inputs smaller than the
keysize,
.Nm
will extract the public RSA key from an SSH formatted key,
generate a 256 bit session key, and then encrypt the session key using the
public RSA key.
The data itself is encrypted using AES-256-CBC mode with said session key.
.Pp
.Nm
will attempt to encrypt the data for each of the public keys found;
the output from
.Nm
consists of several
.Xr uuencode 1 Ns ed
files.
The first is the encrypted message, followed by a SHA-256 HMAC.
The following files are the session key encrypted for
each of the public keys, followed by version
information.
.Nm
will properly extract, decode and use the suitable
session key when the
.Fl d
flag is specified.
.Pp
When decrypting, if no key is specified via the
.Fl k
flag, then
.Nm
will try ~/.ssh/id_rsa.
.Pp
If the message has been tampered with and does not
match the HMAC, then
.Nm
will abort prior to decryption of the input.
.Pp
Since earlier versions of
.Nm
did not support the use of an HMAC, it is possible
that input generated by an older version of
.Nm
is missing the HMAC.
In that case,
.Nm
will print a warning and exit.
If the user wants to proceed and decrypt the data
without the use of an HMAC, then they may set the
.Ar JASS_NO_HMAC
environment variable.
.Sh FINDING PUBKEYS
.Nm
has several ways of finding the public key of the
recipient(s) when encrypting data.
If no public key file is provided via the
.Fl k
flag, then a recipient may also be specified via the
.Fl g
and
.Fl u
flags.
If
.Fl g
was specified,
.Nm
will expand the given group to individual users.
This expansion will include local and LDAP groups (if
LDAP is enabled) as well as GitHub teams (see Section
GITHUB SERVICE).
.Pp
For each user,
.Nm
will first look for that user's ~/.ssh/authorized_keys file, then the file
/etc/ssh/authorized_keys/<user>.
.Pp
If neither of these is readable, and LDAP is configured by setting the
.Ar LDAPSEARCH
environment variable, then it will look for the field designated by the
.Ar LDAPFIELD
environment variable for the given username in LDAP.
.Pp
Alternatively,
.Nm
may look on a GitHub service (either public GitHub, or
an internal GitHub instance) for the public
key associated with the user.
.Pp
Lastly, the user can also specify a separate command
to retrieve the public key of a user via the
.Fl c
flag.
The given command is executed passing a single
recipient name (i.e., a username) and should return a
public ssh keys in OpenSSH's default format, one per line.
.Sh PASS PHRASE ARGUMENTS
.Nm
allows you to specify the passphrase for your private key via one of the
following methods.
Please note that doing so carries certain risks, and you should carefully
evaluate what may be the best option for you.
.Bl -tag -width pass_password_
.It \fBenv:var\fR
Obtain the password from the environment variable \fBvar\fR.
Since the environment of other processes may be visible via e.g.
.Xr ps 1 ,
this option should be used with caution.
.It \fBfile:pathname\fR
The first line of \fBpathname\fR is the password.
\fBpathname\fR need not refer to a regular file: it could for example
refer to a device or named pipe.
Note that standard Unix file access controls should be used to protect
this file.
.It \fBkeychain:name\fR
.Nm
will use the
.Xr security 1
utility to retrieve the password from the macOS keychain.
.It \fBlpass:name\fR
.Nm
will use the LastPass command-line client
.Xr lpass 1
to retrieve the named password.
You should previously have run 'lpass login' for this
to work.
.It \fBop:name\fR
.Nm
will use the 1Password command-line client
.Xr op 1
to retrieve the named password.
.It \fBpass:password\fR
The actual password is \fBpassword\fR.
Since the password is visible to utilities such as
.Xr ps 1
this form should only be used where security is not important.
.El
.Sh GITHUB SERVICE
.Nm
can look for public SSH keys on a GitHub service.
The URL to use for this can be set via the GITHUB_URL
environment variable.
.Pp
If this variable is set,
.Nm
will retrieve keys for users from GitHub as well as
expand groups specified via
.Fl g
as GitHub teams in the format "org/team" or
"numeric-team-id".
.Pp
If you use an internal GitHub instance that requires
authentication, then
.Nm
can make use of the
.Ar GITHUB_API_TOKEN
environment variable.
If set,
.Nm
will use it to set the Basic HTTP Authentication
header 'Authorization' using the current user's
username.
.Pp
When parsing GitHub data,
.Nm
will assume that the result will be JSON data in the
format of:
.Bd -literal -offset -indent
[
 {
   "id": 12345,
   "key": "ssh-rsa AAAAB3NzaC1...",
 },
 ...
]
.Ed
.Sh EXAMPLES
To generate a secret message encrypted with the file 'bobs_pubkey' and
store it in the file 'secret':
.Bd -literal -offset indent
echo "The ostrich has left the savannah." | \\
        jass -e -k bobs_pubkey > secret
.Ed
.Pp
To send a secret message to 'jschauma':
.Bd -literal -offset indent
echo "The lion sleeps." | jass -u jschauma |  \\
        mail -s "Nothing to see here" jschauma
.Ed
.Pp
To decrypt a secret message generated by
.Nm
using the private ssh key found in 'my_privkey':
.Bd -literal -offset indent
jass -d -k my_privkey < secret
.Ed
.Pp
To encrypt the file service.yml for the user 'jschauma':
.Bd -literal -offset indent
jass -u jschauma -f service.yml >service.yml.enc
.Ed
.Pp
To encrypt data for multiple recipients:
.Bd -literal -offset indent
jass -u user1 -u user2 -u user3 <data
.Ed
.Pp
To encrypt data using the internal GitHub instance
with Basic HTTP Auth:
.Bd -literal -offset indent
export GITHUB_URL="https://git.example.com/api/v3"
export GITHUB_API_TOKEN="abcdef0123456789abcdef0123456789abcdef01"
jass -u user1 <data
.Ed
.Pp
If you have a simple SSH Pubkey Store that returns a
list of public keys for a user via a simple HTTP GET
request, you could create a helper script and then use
the
.Fl c
flag as follows:
.Bd -literal -offset indent
cat >fetchkeys <<EOF
#! /bin/sh
curl -s https://internal.example.com/pubkeys/$1
EOF
chmod a+rx fetchkeys
jass -u user1 -c ./fetchkeys <data
.Ed
.Ed
.Sh ENVIRONMENT
When attempting to query LDAP for public keys,
.Nm
will require the variables LDAPFIELD and LDAPSEARCH to be set.
You can either edit the script and set them in there, or export them in
your environment.
.Pp
The following are example values.
.Bl -tag -width GITHUB_API_TOKEN_
.It GITHUB_URL
https://api.github.com/
.It GITHUB_API_TOKEN
abcdef0123456789abcdef0123456789abcdef01
.It JASS_NO_HMAC
feeling_lucky
.It LDAPFIELD
SSHPubkey
.It LDAPSEARCH
ldapsearch -LLLxh ldap.yourdomain.com -b dc=example,dc=com
.El
.Pp
In order to retrieve public SSH keys from your
internal GitHub instance, you could invoke
.Nm
as follows:
.Bd -literal -offset indent
$ export GITHUB_URL="https://git.your.internal.site/api/v3"
$ jass -u jdoe <data
.Ed
.Sh SEE ALSO
.Xr enc 1 ,
.Xr lpass 1 ,
.Xr openssl 1 ,
.Xr rsautl 1 ,
.Xr security 1 ,
.Xr ssh-keygen 1
.Sh BUGS
.Nm
will only allow RSA keys.

CHANGES
* Sat May 21 2022 - jschauma@netmeister.org
- 7.3
-  add support for password managers ('op', 'lpass',
   macOS keychain)

* Fri May 20 2022 - jschauma@netmeister.org
- 7.2
-  finally allow encrypted RSA keys

* Thu May 05 2022 - jschauma@netmeister.org
- 7.1
-  add support for external commands to fetch ssh
   pubkeys ('-c cmd')

* Wed Mar 04 2020 - jschauma@netmeister.org
- 7.0
-  remove support for KEYKEEPER_URL

* Sat Jan 04 2020 - jschauma@netmeister.org
- 6.1
-  add support for GitHub teams
-  change diagnostics to variadic functions

* Wed Oct 23 2019 - jschauma@netmeister.org
- 6.0
-  add support for an HMAC to verify the message
   integrity and authenticity prior to decryption

* Fri May 24 2019 - jschauma@netmeister.org
- 5.3
-  add support for "OPENSSH PRIVATE KEY" format

* Tue Oct 23 2018 - jschauma@netmeister.org
- 5.2
-  fix GitHub URL handling per Issue#19

* Fri Nov 10 2017 - jschauma@netmeister.org
- 5.1
-  switch to using GitHub API for key retrieval
-  allow basic auth via the GITHUB_API_TOKEN
   environment variable

* Thu Sep 28 2017 - jschauma@netmeister.org
- 5.0
-  remove '-G' option; enable use of GitHub via the
   environment; this also allows the use of
   private GitHub servers
-  allow an empty KEYKEEPER_URL or GITHUB_URL to skip
   those services

* Fri Nov 20 2015 - jschauma@netmeister.org
- 4.1
-  be more lenient in detecting the right session key
   data (PR #12 from directionless)
-  improve test harness to allow selective testing

* Fri Aug 28 2015 - jschauma@yahoo-inc.com
- 4.0
-  remove '-K' option; enable use of KeyKeeper via the
   environment
-  accept a KeyKeeper URL from the KEYKEEPER_URL environment
   variable
-  document KeyKeeper usage in manual page
-  add '-l' option to list recipients
-  merge pull request from tmcarr:
   Some basic style fixes following Go's styleguide

* Mon Apr 20 2015 - jschauma@netmeister.org
- 3.2:
-  merge ability to specify passphrase on command-line from @angadsg

* Wed Apr 15 2015 - jschauma@yahoo-inc.com
- 3.1:
-  add support for KeyKeeper keys

* Wed Sep 24 2014 - jschauma@twitter.com
- 3.0.3
-  [SECURITY-14609]: accept encrypted input with spurious whitespace
-  clarify that '-k' can be used for decryption as well

* Fri Sep 12 2014 - jschauma@twitter.com
- 3.0.2
-  [SECURITY-14355]: better detect when input is garbled

* Thu May 15 2014 - jschauma@twitter.com
- 3.0.1
-  [SECURITY-12394]: call /bin/stty instead of relying on the PATH

* Fri Feb 21 2014 - jschauma@twitter.com
- 3.0
-  [SECURITY-11509]: correctly generate fingerprint
   This introduces an incompatibility between version 2.1/2.2 and this
   version, which did not correctly generate a fingerprint for ssh pubkeys
   with a small exponent.

* Thu Jan 30 2014 - jschauma@twitter.com
- 2.2:
-  [SECURITY-11290]: re-introduce backwards compatibility with username-fp
   format
-  sort cap options properly
-  re-add "-f" description into usage

* Mon Jan 27 2014 - jschauma@twitter.com
- 2.1
-  remove dependencies on openssl(1) and ssh-keygen(1) altogether by
   rewriting all encryption/decryption natively in Go
   (this happens to also address [SECURITY-11186])
-  fix tyop in usage

* Tue Dec 31 2013 - jschauma@twitter.com
- 2.0
-  complete rewrite of jass(1) in Go
-  no longer depend on ssh-keygen(1)'s ability to convert ssh keys to
   PKCS8 (as is broken in OS X Mavericks, or unavailable in OpenSSH before
   5.6)

* Fri Dec 13 2013 - jschauma@twitter.com
- 1.8.1
-  better detection of compatibilities
-  add dependency on curl for the rpm

* Thu Dec 12 2013 - jschauma@twitter.com
- 1.8.0
-  add support for fetching pubkeys from GitHub

* Tue Nov 26 2013 - jschauma@twitter.com
- 1.7.2:
-  [SECURITY-10466]: handle keys with options and multiple comments
-  document '-E' in usage

* Wed Sep 25 2013 - jschauma@twitter.com
- 1.7.1:
-  add an option to skip the version check
-  [SECURITY-9720]: if we can't decrypt the session key, suggest to the
   user that they might have entered the wrong passphrase
-  [SECURITY-9755]: export changelog into a separate file

* Fri Aug 30 2013 - jschauma@twitter.com
- 1.7:
-  add '-V' to usage statement
-  allow decryption using a privkey only (ie, extract pubkey from privkey
   instad of relying on the pubkey to exist alongside the privkey)

* Wed Aug 28 2013 - jschauma@twitter.com
- 1.6:

src/jass.sh
#! /bin/sh
#
# Copyright (c) 2013 Twitter, Inc.
#
# Originally written by Jan Schaumann <jschauma@twitter.com> in April 2013.
#
# This script allows you to easily share secrets with other users in your
# organization by way of ssh pubkeys.

# Set these two variables to match your use of LDAP for ssh keys.
#LDAPFIELD="SSHPubkey"
#LDAPSEARCH="ldapsearch -LLLxh ldap.yourdomain.com -b dc=example,dc=com"

### You should not have to edit anything below this line.

set -e

###
### Globals
###

umask 077

VERSION="1.8.1"
CAN_DECRYPT_FROM="1.8 1.7 1.6 1.5 1.4"
CAN_ENCRYPT_FOR="1.8 1.7 1.6 1.5"

PROGNAME="${0##*/}"

ACTION="encrypt"
URLS=""
EXPERT="no"
FILE="/dev/stdin"
GITHUB="https://github.com/<user>.keys"
KEYFILES=""
RECIPIENTS=""
RECIPIENT_GROUPS=""
UNAME="$(uname -s)"
BASE64="base64"

TDIR="$(mktemp -d "${TMPDIR:-/tmp}/jass.XXXX")"
LDAP_KEYFILE="${TDIR}/ldap"
KEYDIR="${TDIR}/pubkeys"
PEMDIR="${TDIR}/pems"
SESSION_KEY="${TDIR}/secret"
URLS_KEYFILE="${TDIR}/urls"

VERBOSITY=0

###
### Functions
###

# Read one ssh key per line from the given file and append it to the
# 'keyfile'.  This will strip any key options and comments, then add the
# given recipient name as a comment to the file.
appendKey() {
	local readonly keyfile="${TDIR}/keyfile"

	local readonly file="${1}"
	local recipient="${2}"

	if [ ! -s "${file}" ]; then
		return
	fi

	oIFS="${IFS}"
	IFS='
'

	# If 'file' and 'recipient' are identical, then we were given a
	# pathname instead of the keys of a user.  We then provide the
	# basename of the path as an identifying comment.
	if [ x"${file}" = x"${recipient}" ]; then
		recipient="${recipient##*/}"
	fi

	verbose "Appending keys for '${recipient}' to keyfile'..." 3

	# SSH keys are more complex than you think.  They may contain
	# leading options (including quoted strings containing spaces) and
	# trailing comments.  All we know -- and care about -- are the
	# actual key.  Since we only support RSA keys, we can ignore the
	# other options; if somebody creates a key with this pattern as a
	# comment, then they can go screw themselves.
	for line in $(sed -n -e 's/.*\(ssh-rsa AAAAB3NzaC1[^ ]*\).*/\1/p' "${file}"); do
		echo "${line} ${recipient}" >> "${keyfile}"
	done

	IFS="${oIFS}"
}


# Let us pretend!  Versions are always either M.m or M.m.t.  We claim that
# tiny version numbers would not introduce any incompatibility issues, so
# we strip that.  This allows us to have tiny version increments without
# having to explicitly list every single tiny version in our compatibility
# list.
checkCompatibility() {
	local greater v v_maj_min
	local pass=0
	local readonly this_major_min=$(echo "${VERSION}" | sed -e 's/\([0-9]*\.[0-9]*\).*/\1/')

	verbose "Checking version compatibility..." 2

	if [ ! -f "${KEYDIR}/version" ]; then
		verbose "Skipping version compatibility test for input without version data." 3
		return
	fi

	for v in $(awk -F: '/CAN_ENCRYPT_FOR/ { print $2 }' "${KEYDIR}/version"); do
		v_maj_min=$(echo ${v} | sed -e 's/\([0-9]*\.[0-9]*\).*/\1/')
		greater=$(echo "${this_major_min} >= ${v_maj_min}" | bc)
		if [ "${greater}" -gt 0 ]; then
			return
		fi
	done

	echo "Incompatible jass format.  Input was generated using:" >&2
	cat "${KEYDIR}/version" >&2
	echo >&2
	echo "This is jass version ${VERSION}."
	exit 1
	# NOTREACHED
}


cleanup() {
	rm -fr "${TDIR}"
}

convertKeyFiles() {
	local pemfile pubkey
	local success=0

	verbose "Converting ssh pubkeys into PKCS8 format..." 2

	for pubkey in "${KEYDIR}"/*; do
		pemfile=$(basename ${pubkey})
		ssh-keygen -P '' -f "${pubkey}" -e -m PKCS8 > "${PEMDIR}/${pemfile}" 2>/dev/null ||	\
			verbose "Unable to convert ssh key '$(basename ${pubkey})' to PKCS8 format."
		if [ -s "${PEMDIR}/${pemfile}" ]; then
			success=1
		fi
	done

	if [ ${success} -eq 0 ]; then
		echo "No valid PKCS8 file could be generated." >&2
		exit 1
	fi
}

# split the single file containing all pubkeys into per-pubkey files
createKeyFiles() {
	local readonly keyfile="${TDIR}/keyfile"
	local found fp recipient type

	found=0

	verbose "Creating per-pubkey files..." 3
	(
		cd "${KEYDIR}"
		grep -v "^#" "${keyfile}" | grep . | sort -u | split -a 6 -l 1 - .key-
	)

	for k in "${KEYDIR}"/.key-*; do
		# We added the recipient as the last field in getPubkeys.
		recipient="$(awk '{print $NF}' "${k}")"
		type=$(ssh-keygen -e -f ${k} | sed -n -e 's/Comment:.*bit \(.*\),.*/\1/p')
		if [ x"${type}" != x"RSA" ]; then
			verbose "Key type of type '${type}' not supported. Skipping." 4
			continue
		fi

		fp=$(getFingerprintFromKey ${k} encrypt)
		if [ -n "${fp}" ]; then
			found=1
			mv ${k} "${KEYDIR}"/${recipient}-${fp}
		else
			verbose "Skipping invalid keyfile '${k}'..." 3
		fi
	done

	if [ ${found} -eq 0 ]; then
		echo "No supported key found." >&2
		exit 1
	fi
}

decrypt() {
	local output="/dev/stdout"
	local outfile skey skeyfp
	local f
	local found=0

	local readonly msgdir="${TDIR}/message"
	local readonly localfp=$(getFingerprintFromKey ${KEYFILES} decrypt)

	if [ -z "${localfp}" ]; then
		exit 1
	fi

	verbose "Decrypting..."

	verbose "Decoding data..." 2
	verbose "Splitting input..." 3
	cat "${FILE}" | ( cd "${TDIR}" && awk '/^begin-base64 / { 		\
							f="data"i++; }		\
						{ print > f; }			\
						/^====/ { close(f); }' )

	if [ $? -gt 0 ]; then
		echo "Invalid input?" >&2
		exit 1
	fi

	verbose "Uudecoding ecnrypted message..." 3
	( mkdir -p "${msgdir}" && cd "${msgdir}" && uudecode ../data0 )
	outfile=$(basename "${msgdir}"/*)

	verbose "Uudecoding all session keys..." 3
	if [ ! -s "${TDIR}"/data1 ]; then
		echo "No encoded session keys found." >&2
		exit 1
	fi
	( cd "${KEYDIR}" && uudecode ../data[1-9]* )

	if [ x"${EXPERT}" = x"no" ]; then
		checkCompatibility
	fi

	verbose "Identifying matching session key..." 3

	for f in "${KEYDIR}"/*; do
		skey=$(basename "${f}")
		skeyfp=$(echo ${skey} | sed -e 's/.*-//')

		if [ x"${skeyfp}" = x"${localfp}" ]; then
			verbose "Using ${skey}..." 4
			found=1
			break
		fi
	done

	if [ ${found} -eq 0 ]; then
		echo "Data was not encrypted for the key in ${KEYFILES}." >&2
		exit 1
	fi

	verbose "Decrypting session key..." 2
	openssl rsautl -inkey ${KEYFILES} -decrypt -in "${KEYDIR}/${skey}" -out "${TDIR}/skey" || {
		echo >&2
		echo "Unable to decrypt session key." >&2
		echo "Are you sure the passphrase unlocks ${KEYFILES}?" >&2
		exit 1
	}

	verbose "Decrypting message using session key..." 2
	openssl enc -d -aes-256-cbc -pass "file:${TDIR}/skey"	\
		-in "${msgdir}/${outfile}" -out "${output}"
}

encrypt() {
	local outfile="message"
	local fp
	local pemfile

	verbose "Encrypting..."

	if [ x"${FILE}" != x"/dev/stdin" ]; then
		if [ ! -r "${FILE}" ]; then
			echo "Unable to read '${FILE}'." >&2
			exit 1
		fi
		outfile=$(basename "${FILE}")
	fi

	getPubkeys
	createKeyFiles
	convertKeyFiles
	generateSessionKey

	verbose "Encrypting data with session key..." 2
	openssl enc -aes-256-cbc -salt -in "${FILE}" -pass "file:${SESSION_KEY}" | \
		uuencode -m "${outfile}"

	verbose "Encrypting session key with pubkeys..." 2
	for pemfile in "${PEMDIR}"/*; do
		fp=$(basename "${pemfile}")
		verbose "Encrypting session key with pubkey ${fp}..." 3
		openssl rsautl -pubin -inkey "${pemfile}" -encrypt -pkcs -in "${SESSION_KEY}" |	\
			uuencode -m ${fp}
	done

	verbose "Encoding version information..." 2
	( echo "VERSION: ${VERSION}";
	  echo "CAN_ENCRYPT_FOR: ${CAN_ENCRYPT_FOR}";
	  echo "CAN_DECRYPT_FROM: ${CAN_DECRYPT_FROM}";
	) | uuencode -m version
}

expandGroup() {
	local readonly group="${1}"
	local all group_users

	verbose "Expanding group '${group}'..." 2

	# Supplementary groups are easy...
	verbose "Expanding group '${group}' by supplementary group membership..." 3
	group_users="$(grep "^${group}:" /etc/group | awk -F: '{print $NF}' | tr ',' ' ')"
	all="${group_users}"
	if [ -n "${all}" ]; then
		verbose "${all}" 4
	fi

	# ...but some users have primary groups...
	gid=$(grep "^${group}:" /etc/group | awk -F: '{print $3}')
	if [ -n "${gid}" ]; then
		verbose "Expanding group '${group}' by primary group membership..." 3
		group_users="$(awk -v gid="${gid}" -F: '{ if ($4 == gid) { print $1; }}' /etc/passwd)"
		if [ -n "${group_users}" ]; then
			all="${group_users} ${all}"
			verbose "${all}" 4
		fi
	fi

	# ...and of course there's LDAP.
	if [ -z "${all}" ]; then
		if ldapVarCheck; then
			verbose "Expanding group '${group}' from LDAP..." 3
			group_users="$(${LDAPSEARCH} cn="${group}" memberUid | grep -v ^dn: | awk '{print $2}')"
			if [ -n "${group_users}" ]; then
				all="${group_users} ${all}"
				verbose "${all}" 4
			fi
		fi
	fi

	RECIPIENTS="${RECIPIENTS} $(echo ${all} | sort -u)"
}

generateSessionKey() {
	verbose "Generating a session key..." 2

	openssl rand 32 -base64 -out "${SESSION_KEY}" 2>/dev/null
}

getFingerprintFromKey() {
	local readonly keyfile="${1}"
	local readonly mode="${2}"
	local fp

	verbose "Extracting fingerprint from '${keyfile}'..." 4

	fp=$(ssh-keygen -l -f "${keyfile}" | awk '{print $2}')

	# ssh-keygen does not yield errors to stderr; if the input is not
	# a valid key, it will print "mumble is not a public key" instead.

	# When decrypting, $keyfile will be a privkey, and we can only get
	# the fingerprint as above if the pubkey happens to be stored
	# alongside the privkey.  (It was worth trying.)
	#
	# Otherwise, we need to extract the pubkey from the privkey, for
	# which we need to prompt for the password.  Unfortunately we
	# can't intercept the password prompt here easily, so the user may
	# end up being prompted again later.

	if [ ! -z "${fp}" -a x"${fp}" = x"is" ]; then
		if [ x"${mode}" = x"encrypt" ]; then
			echo "${keyfile} appears not to be a valid ssh key." >&2
			return
		else
			verbose "Trying to convert '${keyfile}' to a pubkey..." 4
			ssh-keygen -y -f "${keyfile}" >"${TDIR}/pubkey" </dev/tty || {
				echo "Unable to convert ${keyfile} into a pubkey." >&2
				echo "Either the file is not a privkey or the passphrase didn't match." >&2
				return
			}
			fp=$(ssh-keygen -l -f "${TDIR}/pubkey" | awk '{print $2}')
		fi
	fi

	echo "${fp}"
}


# Retrieve all available keys for all recipients and store them in a
# single file, 'keyfile'.
getPubkeys() {
	verbose "Identifying/retrieving ssh pubkeys..." 2

	local found=0
	local etcfile group kfile tildefile
	local readonly keyfile="${TDIR}/keyfile"

	for group in ${RECIPIENT_GROUPS}; do
		expandGroup "${group}"
	done

	if [ -n "${RECIPIENTS}" ]; then
		for r in ${RECIPIENTS}; do
			found=0
			# Tilde expansion happens before variable expansion, so
			# we need to eval echo here.
			tildefile=$(eval echo ~${r}/.ssh/authorized_keys)
			etcfile="/etc/ssh/authorized_keys/${r}"
			for f in "${tildefile}" "${etcfile}"; do
				verbose "Trying '${f}'..." 3
				if [ -r "${f}" ]; then
					appendKey "${f}" "${r}"
					found=1
				fi
			done

			if [ ${found} -eq 0 ]; then
				getPubkeysFromLDAP "${r}" || true
				if [ -s "${LDAP_KEYFILE}" ]; then
					appendKey "${LDAP_KEYFILE}" "${r}"
					found=1
				fi
			fi

			if [ ${found} -eq 0 -a -n "${URLS}" ]; then
				getPubkeysFromURLs "${r}" || true
				if [ -s "${URLS_KEYFILE}" ]; then
					appendKey "${URLS_KEYFILE}" "${r}"
					found=1
				fi
			fi
		done
	fi

	if [ -n "${KEYFILES}" ]; then
		for kfile in ${KEYFILES}; do
			appendKey "${kfile}" "${kfile}"
		done
	fi

	if [ ! -s "${keyfile}" ]; then
		echo "Unable to find any usable pubkeys." >&2
		exit 1
	fi
}

getPubkeysFromLDAP() {
	local readonly user="${1}"
	local readonly oIFS="${IFS}"

	ldapVarCheck || return

	verbose "Trying to get ssh pubkeys for '${user}' from LDAP..." 3

	${LDAPSEARCH} uid="${user}" ${LDAPFIELD}				| \
		grep -v dn: 							| \
		sed -e 's/^ *//' 						| \
		tr -d '\n'							| \
		awk -v F="${LDAPFIELD}" '{ gsub(F ":* ", "\n"); print; }'	| \
		sed -e 's/^%.*//'						| \
		grep . > "${LDAP_KEYFILE}" || {
			verbose "Unable to retrieve ssh pubkey for '${user}' from LDAP."
			return
	}

	# Keys may be base64, in which case (we assume) it will not
	# contain any spaces, otherwise required for a valid ssh key.

	IFS="
"

	for line in $(cat "${LDAP_KEYFILE}"); do
		if echo "${line}" | grep -q ' '; then
			echo "${line}" >> "${LDAP_KEYFILE}.new"
		else
			echo "${line}" | eval ${BASE64}				| \
				sed -e 's/^%.*//'				| \
				grep . > "${LDAP_KEYFILE}.tmp" || {
					echo "Ignoring unusable data in LDAP for '${user}'." >&2
					continue
				}
			cat "${LDAP_KEYFILE}.tmp" >> "${LDAP_KEYFILE}.new"
			rm "${LDAP_KEYFILE}.tmp"
		fi
	done

	if [ -s "${LDAP_KEYFILE}.new" ]; then
		mv "${LDAP_KEYFILE}.new" "${LDAP_KEYFILE}"
	fi

	IFS="${oIFS}"

	if [ ! -s "${LDAP_KEYFILE}" ]; then
		echo "Unable to retrieve ssh pubkey for '${user}' from LDAP." >&2
		return
	fi
}

getPubkeysFromURLs() {
	local readonly user="${1}"
	local url

	verbose "Trying to get ssh pubkeys for '${user}' from URLs..." 3
	rm -f "${URLS_KEYFILE}"
	for url in ${URLS}; do
		url=$(echo "${url}" | sed -e "s/<user>/${user}/g")
		verbose "Trying to get '${url}'..." 4
		curl -s "${url}" >> "${URLS_KEYFILE}" || {
			echo "Unable to curl '${url}'." >&2
		}
	done
}


usage() {
	cat <<EOH
Usage: ${PROGNAME} [-EVdehv] [-f file] [-k key] [-u user]
        -E       expert mode; skip certain tests
	-V       print version information and exit
	-d       decrypt
	-e       encrypt (default)
	-f file  encrypt/decrypt this file
	-h       print this help and exit
	-k key   encrypt using this public key file
	-u user  encrypt for this user
	-v       be verbose
EOH
}

ldapVarCheck() {
	if [ -z "${LDAPSEARCH}" -o -z "${LDAPFIELD}" ]; then
		verbose "Use of LDAP not configured."
		verbose "Please set the LDAPSEARCH and LDAPFIELD variables near the top of" 2
		verbose "this script or in your environment." 2
		return 1
	fi
	return 0
}


# purpose : check that all important variables are set correctly
# input   : none
# output  : none, globals may have been modified, may abort on error

varCheck() {
	local file kfiles n

	verbose "Checking that all variables look ok..."
	n=0
	if [ -n "${KEYFILES}" ]; then
		for file in ${KEYFILES}; do
			if [ ! -r "${file}" ]; then
				echo "Skipping unreadable key file '${file}'." >&2
			else
				kfiles="${file} ${kfiles}"
				n=$(( ${n} + 1 ))
			fi
		done
		if [ -z "${kfiles}" ]; then
			echo "No usable key file found." >&2
			exit 1
		fi
		KEYFILES="${kfiles}"
	fi

	if [ x"${ACTION}" = x"encrypt" ]; then
		if [ -z "${KEYFILES}" -a -z "${RECIPIENTS}" -a -z "${RECIPIENT_GROUPS}" ]; then
			echo "You need to provide either a key file, a group, or a username." >&2
			exit 1
		fi
	else
		if [ -n "${RECIPIENT_GROUPS}" ]; then
			echo "You can't specify a group when decrypting." >&2
			exit 1
		fi
		if [ -n "${RECIPIENTS}" ]; then
			echo "You can't specify a user when decrypting." >&2
			exit 1
		fi

		if [ ${n} -gt 1 ]; then
			echo "Please only specify a single key file when decrypting." >&2
			exit 1
		fi

		if [ -z "${KEYFILES}" ]; then
			verbose "No key specified, trying ~/.ssh/id_rsa..." 2
			KEYFILES="${HOME}/.ssh/id_rsa"
			if [ ! -r "${KEYFILES}" ]; then
				echo "No key found.  Please use '-k'." >&2
				exit 1
			fi
		fi
	fi

	if [ x"${UNAME}" = x"Linux" ]; then
		BASE64="base64 -d -i"
	else
		BASE64="base64 -D"
	fi
}


verbose() {
	local msg="${1}"
	local level="${2:-1}"
	local i=0

	if [ ${level} -le ${VERBOSITY} ]; then
		while [ ${i} -lt ${level} ]; do
			printf "=" >&2
			i=$(( ${i} + 1 ))
		done
		echo "> ${msg}" >&2
	fi
}

###
### Main
###

trap 'cleanup' 0

while getopts 'EGVdef:g:hk:u:v' opt; do
	case ${opt} in
		E)
			EXPERT="yes"
		;;
		G)
			URLS="${URLS} ${GITHUB}"
		;;
		V)
			echo "jass version ${VERSION}"
			echo "Able to decrypt input generated with jass versions: ${CAN_DECRYPT_FROM}"
			echo "Able to encrypt data for jass versions: ${CAN_ENCRYPT_FOR}"
			exit 0
			# NOTREACHED
		;;
		d)
			ACTION="decrypt"
		;;
		e)
			ACTION="encrypt"
		;;
		f)
			if [ x"${OPTARG}" != x"-" ]; then
				FILE="${OPTARG}"
			fi
		;;
		g)
			RECIPIENT_GROUPS="${OPTARG} ${RECIPIENT_GROUPS}"
		;;
		h\?)
			usage
			exit 0
			# NOTREACHED
		;;
		k)
			KEYFILES="${OPTARG} ${KEYFILES}"
		;;
		u)
			RECIPIENTS="${OPTARG} ${RECIPIENTS}"
		;;
		v)
			VERBOSITY=$(( ${VERBOSITY} + 1 ))
		;;
		*)
			usage
			exit 1
			# NOTREACHED
		;;
	esac
done
shift $(($OPTIND - 1))

if [ $# -ne 0 ]; then
	usage
	exit 1
	# NOTREACHED
fi

varCheck

mkdir -p "${KEYDIR}" "${PEMDIR}"

${ACTION}

exit 0
